Drag and Droppable #128

ppelayo1 opened the pull request:
### :scroll: Description

  This pull request solves the same problem as the Ic/admin table one. Due to the number of changes I decided to just open a new request with a new branch, and will be closing the other pull request.

- Describe the general shape of this PR (new feature? refactor? bug fix? one-line change?)
<!--- Add your answer here -->
  Implemented DND kit for support in sorting agenda items

- Describe what changes are being made
<!--- Add your answer here -->
  AgendaView, AgendaGroup(renamed AgendaGroups), and AgendaItem were heavily modified to support the swapping of the agenda groups as well as the items within the agenda groups. The items can swap between containers as well as within containers.

  AgendaGroup  renamed to AgendaGroups was changed from handling a single group to handling all groups. This was necessary for the sortable context to be able to handle the sorting of the group containers themselves.

  AgendaItem had to be modified to include a input checkmark that would overlay on top of the normal one. This had to be done so that the box was checkable. dragOverlay component from DND kit would not allow the checking of the check mark, so I made that overlay directly above it, outside of the dragOverlay.
The overlay checkmark is hidden when the items are being dragged around, and only appear when nothing is being dragged.
AgendaItem.scss holds the CSS classes that control the location of the overlay checkmark and its position.

- Describe why these changes are being made
 To address issue Agenda Items: Drag and Droppable #128 
  I made the agenda groups themselves sortable as well. This can be removed if needed to do so.

- List the use cases and edge cases relevant to this PR
Use case would be for the moving of agenda items from one group to another, or between the same group.
Edge case would be large number of agenda items and groups, as well as quickly moving agenda items between groups open and closed.

- Describe how errors will be handled. How will we know if this code breaks in production


- Describe any external libraries/dependencies added or removed in this PR
Added dnd-kit/core and dnd-kit/sortable

- Describe any security risks are there regarding this change


- Describe how you tested these changes
I created roughly 5 agenda groups and 5 items for each group. 2 were completed and full of completed items, 2 pending items, 1 was empty with a in progress status.
I moved the items around manually between groups. Tested with and without "show completed items" checked. Made sure to move between containers, fill containers with all items. Move items back. I checked and unchecked the items. Moved between containers checked and unchecked items.

- Link to relevant external documentation
<!--- Add your answer here -->
https://dndkit.com/


--------------------------
### :clipboard: Mandatory Checklist

- [ ] Linked to the Github Issues being addressed using the right sidebar :arrow_right:
- [ ] Have you discussed these changes with the project leader(s)?
- [ ] Do all variable and function names communicate what they do?
- [ ] Were all the changes commented and / or documented?
- [ ] Is the PR the right size? (If the PR is too large to review, it might be good to break it up into multiple PRs.)
- [ ] Does all work in progress, temporary, or debugger code have a TODO comment with links to Github issues?
- [ ] *If you changed the user interface, did you add before and after screenshots to below?*

--------------------------
### :framed_picture: Screenshots and Screen Recordings

#### Before
<!--- Add before screenshots here -->


#### After
<!--- Add after screenshots here -->


--------------------------
### :blue_book: Glossary
- PR = Pull Request

JMStudiosJoe reviewed (dismissed):
lgtm

anna-kartynnik commented on frontend/src/components/MeetingView/AgendaView/AgendaGroups.jsx:
```diff
@@ -0,0 +1,154 @@
+        <div style={style} ref={setNodeRef}>
+          {agendaGroup.items.map((item) => (
+            <AgendaItem
+              id={item.id}
```
Is it required to have `id` as a separate property? Since you pass `item`, you have access to its `id` via `item.id`.

ppelayo1 replied:
You are right, a id doesn't need to be passed there. I still have to pass it to RenderedAgendaItem in AgendaItem.jsx from the AgendaItem so that the forwardRef can recieve it.

I went ahead and removed the id=item.id from line 139 and used item in AgendaItem.

anna-kartynnik commented on frontend/src/components/MeetingView/AgendaView/AgendaItem.jsx:
```diff
@@ -45,19 +51,76 @@ const itemLinks = [
+}, ref) => {
   const handleCheck = (evt) => {
-    if (evt.target) {
+    if (evt.target && typeof handleSelection !== 'undefined') {
```
In what cases `handleSelection` is `undefined`?

ppelayo1 replied:
You are right, didn't need that. Went ahead and removed it.

anna-kartynnik commented on frontend/src/components/MeetingView/AgendaView/AgendaView.jsx:
```diff
@@ -25,12 +39,81 @@ import MeetingItemStates from '../../../constants/MeetingItemStates';
+  const itemsWithNoParent = items.filter((item) => item.parent_meeting_item_id === null);
+  const itemsWithParent = items.filter((item) => item.parent_meeting_item_id !== null);
+
+  const agendaGroups = [];
```
Are there any particular reasons you changed dictionary structure to array here? If you have dictionary, next part with `itemsWithParent.forEach(...)` runs in linear time. Now it runs in quadratic time.

ppelayo1 replied:
I changed it as I found working with an array directly was easier. I didn't think about the performance impact at the time. I can work at converting it back to a dictionary format if you want me to.

anna-kartynnik replied:
I see. I think in this case we should be fine since we don't expect a lot of items in one meeting.

By the way, do you use the fact that `agendaGroups` are ordered now? Is it important for drag and drop functionality?

ppelayo1 replied:
The order does matter for the DND framework. As an Object there was no order. I been messing around with the code this morning. One option to get around being an object is a new state variable that holds the render order.

So I think I could make it work with an object, but if the performance hit is negligible it will save me the headache of redoing a bunch of the code.

anna-kartynnik replied:
That was my guess: that the order now is really important. We will leave as it is now, no need to introduce more complexity in this case.

anna-kartynnik reviewed (changes requested):
Looks good in general, there are a few small questions below.

ppelayo1 commented on frontend/src/components/MeetingView/AgendaView/AgendaGroups.jsx:
```diff
@@ -0,0 +1,154 @@
+          />
+        </AccordionItem>
+      ))}
+      ;
```
You are right, that shouldn't be there. Went ahead and removed that semi-colon.

ppelayo1 commented:
I also no longer needed the dragOverlayActive variable in the components, it was needed before I implemented a distance requirement on the drag. So I went ahead and removed it.

anna-kartynnik reviewed (approved):
@ppelayo1 do you think it would be possible to split the code of `AgendaView` component into several components? Not in this PR, but maybe in the future? Right now it's quite a big component.

ppelayo1 commented:
I am sure I could find a way to further split code up in the future.
